cirq/ops/pauli.py

```diff
@@ -11,73 +11,56 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+import abc
+from typing import Any, Union, overload, TYPE_CHECKING
 
-from typing import Optional, Union, overload, TYPE_CHECKING
-
-import numpy as np
-
-from cirq import protocols, value
-from cirq.ops import common_gates
-from cirq.type_workarounds import NotImplementedType
+from cirq import value
+from cirq.ops import common_gates, eigen_gate
 
 if TYPE_CHECKING:
     # pylint: disable=unused-import
     from typing import Tuple
 
 
-@value.value_equality
-class Pauli:
-    """Represents the X, Y, or Z axis of the Bloch sphere."""
-    X = None  # type: Pauli
-    Y = None  # type: Pauli
-    Z = None  # type: Pauli
-    XYZ = None  # type: Tuple[Pauli, Pauli, Pauli]
+class Pauli(eigen_gate.EigenGate, metaclass=abc.ABCMeta):
+    """Represents the Pauli gates.
+
+    This is an abstract class with no public subclasses. The only instances
+    of private subclasses are the X, Y, or Z Pauli gates defined below.
+    """
+    _XYZ = None  # type: Tuple[Pauli, Pauli, Pauli]
 
-    def __init__(self, *, _index: int, _name: str) -> None:
+    @staticmethod
+    def by_index(index: int) -> 'Pauli':
```

Comment:
Do these gates have specialized simulation methods? It's important that X and Z go fast. If not, open an issue to resolve.

Reply:
Yes. Every object that we return here is either cirq.X, cirq.Y or cirq.Z and therefore `isinstance` of XPowGate, YPowGate or ZPowGate. All `isinstance`-based specializations for these types trigger for all Pauli gates appropriately.

I've added a unit test to verify `isinstance` behavior.